File: src/main/groovy/org/mulima/api/audio/AudioFormat.java
package org.mulima.api.audio;

import java.io.File;

import org.springframework.util.StringUtils;

public enum AudioFormat {
	WAVE("wav"),
	FLAC("flac"),
	VORBIS("ogg"),
	AAC("m4a"),
	MP3("mp3");
	
	private final String ext;
	
	/**
	 * Constructs an audio file type from an extension.
	 * @param ext the extension
	 */
	private AudioFormat(String ext) {
		this.ext = ext;
	}

	/**
	 * Gets the file extension used for this type.
	 * @return file extension
	 */
	public String getExtension() {
		return ext;
	}
	
	/**
	 * Tests a file to see if it is of the
	 * same type.  Uses the file extension.
	 * @param file file to test.
	 * @return <code>true</code> if of the same type, <code>false</code> otherwise
	 */
	public boolean isFormatOf(File file) {
		String extension = StringUtils.getFilenameExtension(file.getAbsolutePath());
		return this.getExtension().equals(extension);
	}
	
	/**
	 * Gets the file type of a given file.
	 * @param file the file to get the type of
	 * @return the type of the file
	 */
	public static AudioFormat valueOf(File file) {
		String extension = StringUtils.getFilenameExtension(file.getAbsolutePath());
		for (AudioFormat type : AudioFormat.values()) {
			if (type.getExtension().equals(extension)) {
				return type;
			}
		}
		throw new IllegalArgumentException("No type with extension \"" + extension + "\" exists.");
	}
}

File: src/main/scala/model/AudioFormat.scala
package org.ajoberstar.mulima
package model

import java.nio.file.Path

object AudioFormat {
	private val extensionsByFormat: Map[AudioFormat, Set[String]] = Map(
		WAVE -> Set("wav", "wave"),
		FLAC -> Set("flac"),
		VORBIS -> Set("ogg"),
		AAC -> Set("m4a"),
		MP3 -> Set("mp3")
	)

	def apply(extension: String): Option[AudioFormat] = {
		val lowerExtension = extension.toLowerCase
		extensionsByFormat.find { case (format, extensions) =>
			extensions.exists(_ == lowerExtension)
		}.map { case (format, extensions) =>
			format
		}
	}

	def primaryExtension(format: AudioFormat): String = extensionsByFormat(format).head

	def isLossy(format: AudioFormat): Boolean = format match {
		case WAVE => false
		case FLAC => false
		case _ => true
	}
}

sealed trait AudioFormat {
	def primaryExtension: String = AudioFormat.primaryExtension(this)
}
case object WAVE extends AudioFormat
case object FLAC extends AudioFormat
case object VORBIS extends AudioFormat
case object AAC extends AudioFormat
case object MP3 extends AudioFormat
